### ISSUE TITLE
return value of dispatch(addListener) using a thunk's dispatch isn't considered a function

### ISSUE THREAD

**Issue by abrouaux** [reduxjs/redux-toolkit#3139]
Version used: 1.9.2

Implementing a "wait for some action in a thunk pattern", I noticed that the return value of the `dispatch(addListener)` call isn't considered to be a function by typescript, so I was forced to "as any" the return value to unsubscribe.

I assume the typing difficulties that is explained in https://blog.isquaredsoftware.com/2022/03/designing-rtk-listener-middleware/ needs to happen on the dispatch from the thunk - it s rems to be a different typing definition.

Potential bonus: would it be possible to unify things & get the same listener API functions (take, fork, etc) in thunks? That would be quite helpful in many scenarii.
ATM I'm simulating a take via a specific thunk (which exhibit the unsubscribe issue), like so:

```
const awaitPlayerCreationThunk = createAsyncThunk<void, void, { state: PlayerState }>(
  "thunk/awaitPlayerCreation",
  async (_, { dispatch, getState }) => {
    const player = (await import("./PlayerInstances")).playerInstanceFromId(getState().storeId)?.player;
    if (player) return;

    await new Promise<void>(resolve => {
      const unsubscribe = dispatch(
        addListener({
          actionCreator: playerThunks.createInstance.fulfilled,
          effect: () => {
            // typescript issue with RTKQuery
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (unsubscribe as any)();
            resolve();
          },
        })
      );
    });
  }
);
```

**Comment by phryneas** [comment]
That dispatching `addListener` returns a function is a details of the listener middleware - `createAsyncThunk` does not know if you have the listenerMiddleware installed or not, so it cannot do that by default.

You can create a custom-typed version of `createAsyncThunk` using
```ts
const typedCAT = createAsyncThunk.withTypes<{dispatch: AppDispatch}>();
```

That should then be aware of that.

**Comment by EskiMojo14** [comment]
but also, the listener effect receives [unsubscribe](https://redux-toolkit.js.org/api/createListenerMiddleware#listener-subscription-management) as a property of the listenerApi object (second argument provided), so you don't need to do this in the first place.

**Comment by markerikson** [comment]
All that said, this feels like a likely misuse of `createAsyncThunk` to me. Can you give a specific example of where and how you want to wait for this value to be added?

**Comment by abrouaux** [comment]
thanks @EskiMojo14 ! indeed, I didn't realize unsubscribe was in the api object.

@markerikson ok I wasn't thinking about the middleware aspect, I see how it wouldn't make sense now.

But then I'm wondering, isn't the thunk middleware a subset of the listenerMiddleware from a functional standpoint? It seems the listenerMiddleware has a much larger functional scope.

Our usage of thunks is not exactly simple ... we use a video player (that we don't own) with its own async lifecycle & behavior. And which is definitely not React-compatible.

Therefore, I've put in place a redux wrapper around it, to decouple the app from the actual player. The store is mainly about the player internal state at any time.

On one side, the player is dispatching slice actions to update the state for the app to consume.
On the other side, the app is dispatching async thunks to act on the player instance (most of its functions are async).

It works pretty well, I can see state changes & actions in the redux devTools, it's pretty convenient.
The only problem I was having is at creation time. I have two concurrent thunks which aren't dispatched from the same place.
The first is a "createInstance" thunk, which is dispatched when the DOM which will hold the video element is ready (mounted).
The second is a "load" thunk, which is dispatched by the app, to load something into the player.

Given they are concurrent, I need to make sure that if load is dispatched when createInstance hasn't, it waits for its completion, so that from the app's perspective, once I've awaited for the dispatch(load thunk) I'm sure everything is set.

Hence the piece of code I shared initially.

ATM my understanding is that the listenerMiddleware is all about starting async side effect(s) from sync actions. So I guess I could define a lot of trigger actions which only purpose would be to start effects in the listenerMiddleware - and not use the thunks at all anymore? 
But then I don't have the pending/fulfilled/rejected automatic actions from the thunkMiddleware...

What I feel I'm missing is how one would simply use the listenerMiddleware to start an async effect without depending on an action - if at all possible.

**Comment by markerikson** [comment]
> isn't the thunk middleware a subset of the listenerMiddleware from a functional standpoint

Not exactly.  They have some similarities in that they both give you a way to run arbitrary sync or async logic, with access to `dispatch` and `getState`.  But, the use cases are very different.

Conceptually, thunks are for "imperative" logic: "go do this thing, right now, because I'm calling the function to make it happen immediately".  This happens _once_ per time that you explicitly dispatch the thunk.

Listeners are for "reactive" logic: "wait for this action to be dispatched or this state to change, _then_ go run this logic".  This can happen _many_ times based on your conditions, and you can even have many instances of a given listener running simultaneously.

Sagas have always made a lot of use of what I personally refer to as "signal actions" - actions that are dispatched solely to trigger reactive behavior, without any intended state updates.  Listeners are conceptually similar to sagas as a whole, so yes, you could definitely dispatch actions just to kick off listener behavior.

> how one would simply use the listenerMiddleware to start an async effect without depending on an action

At the moment, you can't, exactly. The listener checks are _always_ run after each dispatched action is processed.  We do have a hacky technique over in https://github.com/reduxjs/redux-toolkit/issues/2435 that sort of lets you start a one-shot listener and get back a promise that resolves when the condition is met, but not sure if that's relevant here.  There's also a hypothetical suggestion that we could maybe execute immediately with a fake `type` field.

Could you put together some kind of miniature running example as a sandbox or repo, so that I can better understand what these two thunks are intended to do and how they relate timing-wise?

**Comment by abrouaux** [comment]
Thanks for the explanation. I understand where the whole thunk vs listener middleware come from - their history - and therefore the imperative vs reactive underlying concepts.

Yet, looking at the `createAsyncThunk` code, I'm wondering if one could write a similar function which would have the exact same use, but would be handled by the listenerMiddleware ecosystem.
From what I see, it would just be matter of defining the same set of actions - pending, fulfilled & rejected, then:
- put in place a one-shot listener
- dispatch the pending action
- let the listener run, unsubscribe
- then depending on the outcome of the effect/thunk, dispatch either fulfilled or rejected

That way, the effect/thunk would have access not only to dispatch & gestate, but also to the other extra listenerMiddleware api tools like take, fork, etc.
Which in turn might make the thunkMiddleware redundant?

Am I missing something which would prevent such a thing?

**Comment by markerikson** [comment]
@abrouaux : I still don't think I actually understand the use case or what you're trying to solve.

Could you write some fictional code usage examples and explain what you want here?

**Comment by birnam** [comment]
I'd like to add something to help Google driven visitors. I came across this thread when I wasn't getting the expected 'unsubscribe' return type when adding a listener -- it wasn't a function, and also my listener didn't seem to work.

_Make sure you've also added the listenerMiddleware to the store configuration!_ Event if you're only using listeners on a component level (dynamically added). Once I did this, the return value for dispatch(addListener(...)) changed to the expected unsubscribe function.
